crates/ir-generated/src/emit/register_allocator.rs

```diff
@@ -0,0 +1,94 @@
+use std::{
+    mem,
+    ops::{Deref, Range},
+};
+
+use rokugo_ir::register::{self, Register, RegisterId};
+
+use crate::errors::register::RegisterAllocationError;
+
+pub struct RegisterDropGuard<T: Register> {
+    register: T,
+}
+
+impl<T: Register> RegisterDropGuard<T> {
+    pub(crate) fn new(register: T) -> Self {
+        Self { register }
+    }
+}
+
+impl<T: Register> Deref for RegisterDropGuard<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.register
+    }
+}
+
+#[cfg(debug_assertions)]
+impl<T: Register> Drop for RegisterDropGuard<T> {
```

Comment:
Similar remark about using constraints in `impl` clauses. I'd prefer if we had used `where` for those, except for `Send`, `Sync` and `?Sized`, as they tend to be quite important general API usage constraints. Don't treat this as a hard requirement for passing the review though, it's just code style.